cabal-install/Distribution/Client/Outdated.hs

```diff
@@ -0,0 +1,134 @@
+import Distribution.Types.ComponentRequestedSpec
+       (defaultComponentRequestedSpec)
+import Distribution.Types.Dependency
+       (Dependency(..), depPkgName, simplifyDependency)
+import Distribution.Verbosity                        (Verbosity)
+import Distribution.Version
+       (Version, UpperBound(..), asVersionIntervals, majorBoundVersion)
+
+import qualified Data.Set as S
+import System.Directory                              (getCurrentDirectory)
+import System.Exit                                   (exitFailure)
+
+-- | Entry point for the 'outdated' command.
+outdated :: Verbosity -> OutdatedFlags -> RepoContext
+         -> Compiler -> Platform
+         -> IO ()
+outdated verbosity outdatedFlags repoContext comp platform = do
+  sourcePkgDb <- IndexUtils.getSourcePackages verbosity repoContext
+  let freeze   = fromFlagOrDefault False (outdatedFreeze outdatedFlags)
+      exitCode = fromFlagOrDefault False (outdatedExitCode outdatedFlags)
+      ignore   = S.fromList (outdatedIgnore outdatedFlags)
+      minor    = S.fromList (outdatedMinor outdatedFlags)
+      pkgIndex = packageIndex sourcePkgDb
+  deps <- if freeze
+          then depsFromFreezeFile verbosity
+          else depsFromPkgDesc    verbosity comp platform
+  let outdatedDeps = listOutdated deps pkgIndex ignore minor
+  notice verbosity ("Outdated dependencies: "
+                     ++ intercalate ", "
+                        (map (\(d, v) -> display d
+                               ++ " (latest: " ++ display v ++ ")") outdatedDeps))
+  if (exitCode && (not . null $ outdatedDeps))
+    then exitFailure
+    else return ()
+
+-- | Read the list of dependencies from the freeze file.
+depsFromFreezeFile :: Verbosity -> IO [Dependency]
+depsFromFreezeFile verbosity = do
+  cwd        <- getCurrentDirectory
+  userConfig <- loadUserConfig verbosity cwd Nothing
+  let ucnstrs = map fst . configExConstraints . savedConfigureExFlags $ userConfig
+      deps    = mapMaybe (packageConstraintToDependency . userToPackageConstraint)
+                ucnstrs
+  debug verbosity "Reading the list of dependencies from the freeze file"
+  return deps
+
+-- | Read the list of dependencies from the package description.
+depsFromPkgDesc :: Verbosity -> Compiler  -> Platform -> IO [Dependency]
+depsFromPkgDesc verbosity comp platform = do
+  cwd  <- getCurrentDirectory
+  path <- tryFindPackageDesc cwd
+  gpd  <- readPackageDescription verbosity path
+  let cinfo = compilerInfo comp
+      epd = finalizePD [] defaultComponentRequestedSpec
+            (const True) platform cinfo [] gpd
+  case epd of
+    Left _        -> die "finalizePD failed"
+    Right (pd, _) -> do
+      let bd = buildDepends pd
+      debug verbosity
+        "Reading the list of dependencies from the package description"
+      return bd
+
+-- | Find all outdated dependencies.
+listOutdated :: [Dependency] -> PackageIndex UnresolvedSourcePackage
+              -> S.Set PackageName -> S.Set PackageName
```

Comment:
A set of packages to ignore and a set of packages for which major version bumps should be ignored.